File: evals/run_evaluation.py
#!/usr/bin/env python3
"""
Main evaluation script for MCP tool calling.
"""

import json
import os
import sys
from copy import deepcopy
from pathlib import Path
from typing import Any, Callable

from anthropic import Anthropic
from openai import OpenAI
from phoenix import Client as PhoenixClient
from phoenix.experiments import run_experiment
from phoenix.experiments.types import Example

from config import MODELS_TO_EVALUATE, SYSTEM_PROMPT, validate_env_vars, DATASET_NAME, PASS_THRESHOLD


def load_tools() -> list[dict[str, Any]]:
    """Load current tool definitions from tools.json."""
    tools_path = Path(__file__).parent.parent / 'evals' / 'tools.json'

    if not tools_path.exists():
        print(f'Error: tools.json not found at {tools_path}')
        print("Run 'npm run evals:export-tools' first to export current tool definitions")
        sys.exit(1)

    with open(tools_path, 'r') as f:
        return json.load(f)

def transform_tools_to_openai_format(tools):
    """ Transforms the tools to the OpenAI format."""
    return [
        {
            "type": "function",
            "function": {
                "name": tool["name"],
                "description": tool["description"],
                "parameters": tool["inputSchema"],
            },
        }
        for tool in tools
    ]

def transform_tools_to_antrophic_format(tools):
    """ Transforms the tools to the Antrophic format."""
    t = deepcopy(tools)
    for tool_ in t:
        tool_["input_schema"] = tool_.pop("inputSchema")
    return t


def create_openai_task(model_name: str, tools: list[dict[str, Any]]) -> Callable[[Example], list[str]]:
    """Create OpenAI task function with captured metadata."""
    def task(example: Example) -> list[str]:
        client = OpenAI()
        messages = [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": example.input.get("question")}
        ]

        response = client.chat.completions.create(
            model=model_name,
            messages=messages,
            tools=transform_tools_to_openai_format(tools),
        )
        tool_calls = []
        print(example.input.get('question'), response.choices[0].message)
        if response.choices[0].message.tool_calls:
            tool_calls.append(response.choices[0].message.tool_calls[0].function.name)
        return tool_calls

    return task

def create_anthropic_task(model_name: str, tools: list[dict[str, Any]]) -> Callable[[Example], list[str]]:
    """Create Anthropic task function with captured metadata."""
    def task(example: Example) -> list[str]:
        anthropic_client = Anthropic(timeout=10)

        response = anthropic_client.messages.create(
            model=model_name,
            system=SYSTEM_PROMPT,
            messages=[{"role": "user","content": example.input.get("question")}],
            tools=transform_tools_to_antrophic_format(tools),
            max_tokens=2048,
        )

        tool_calls = []
        print(example.input.get('question'), response.content)
        for content in response.content:
            if content.type == 'tool_use':
                tool_calls.append(content.name)
        return tool_calls

    return task


def tools_match(expected: dict, output: list[str]) -> bool:
    """Check if expected tools match actual output."""
    tool_calls = expected.get('tool_calls', '')
    expected_tools = tool_calls.split(', ') if tool_calls else []
    expected_tools = [tool.strip() for tool in expected_tools if tool.strip()]

    return sorted(expected_tools) == sorted(output)


def main():
    """Main evaluation function."""
    print('Starting MCP tool calling evaluation')

    # Validate environment variables
    if not validate_env_vars():
        sys.exit(1)

    # Load tools and dataset info
    tools = load_tools()

    print(f'Loaded {len(tools)} tools')

    # Initialize Phoenix client
    phoenix_client = PhoenixClient()

    # Get dataset
    try:
        dataset = phoenix_client.get_dataset(name=DATASET_NAME)
    except Exception as e:
        print(f'Error loading dataset: {e}')
        sys.exit(1)

    print(f'Loaded dataset "{DATASET_NAME}" with ID: {dataset.id}')
    # Results storage
    results = []

    # Run evaluations for each model
    for model_name in MODELS_TO_EVALUATE:
        print(f'\nEvaluating model: {model_name}')

        # Initialize variables
        accuracy = 0.0
        correct_cases = 0
        total_cases = 0
        experiment_id = None
        error = None

        if model_name.startswith("gpt"):
            task_fnc = create_openai_task(model_name, tools)
        elif model_name.startswith("claude"):
            task_fnc = create_anthropic_task(model_name, tools)
        else:
            print(f'Unknown model type: {model_name}, skipping')
            results.append(
                {'model_name': model_name, 'accuracy': 0.0, 'correct': 0, 'total': 0, 'error': 'Unknown model type'}
            )
            continue

        # Run experiment
        experiment_name = f'MCP tool calling eval {model_name}'
        experiment_description = f'Evaluation of {model_name} on MCP tool calling'

        # try:
        experiment = run_experiment(
            dataset=dataset,
            task=task_fnc,
            evaluators=[tools_match],
            experiment_name=experiment_name,
            experiment_description=experiment_description,
        )

        # Get evaluations and calculate accuracy
        evaluations_df = experiment.get_evaluations()
        total_cases = len(evaluations_df)
        correct_cases = len(evaluations_df[evaluations_df['score'] > 0.5])
        accuracy = correct_cases / total_cases if total_cases > 0 else 0
        experiment_id = experiment.id

        print(f'{model_name}: {accuracy:.1%} ({correct_cases}/{total_cases})')

        # Print sample of evaluations for debugging
        if evaluations_df is not None and len(evaluations_df) > 0:
            print(f'Sample evaluation results:')
            print(evaluations_df[['score', 'label', 'output', 'expected']].to_string())

        # except Exception as e:
        #     print(f'Error evaluating {model_name}: {e}')
        #     error = str(e)

        # Store results
        results.append({
            'model': model_name,
            'accuracy': accuracy,
            'correct': correct_cases,
            'total': total_cases,
            'experiment_id': experiment_id,
            'error': error,
        })

    # Check if all models meet the pass threshold
    all_passed = all(r.get('accuracy', 0) >= PASS_THRESHOLD for r in results if r.get('error') is None)

    print(f'\nPass threshold: {PASS_THRESHOLD:.1%}')
    if all_passed:
        print('✅ All models passed the threshold')
    else:
        print('❌ Some models failed to meet the threshold')

    return 0 if all_passed else 1


if __name__ == '__main__':

    from dotenv import load_dotenv
    load_dotenv()  # Load environment variables from .env file if present

    exit_code = main()
    sys.exit(exit_code)

File: evals/run-evaluation.ts
#!/usr/bin/env tsx
/**
 * Main evaluation script for MCP tool calling (TypeScript version).
 */

import { createClient } from '@arizeai/phoenix-client';
// eslint-disable-next-line import/extensions
import { getDatasetInfo } from '@arizeai/phoenix-client/datasets';
// eslint-disable-next-line import/extensions
import { asEvaluator, runExperiment } from '@arizeai/phoenix-client/experiments';
import type { ExperimentEvaluationRun, ExperimentTask } from '@arizeai/phoenix-client/types/experiments';
import dotenv from 'dotenv';
import yargs from 'yargs';
// eslint-disable-next-line import/extensions
import { hideBin } from 'yargs/helpers';

import log from '@apify/log';

import {
    loadTools,
    createOpenRouterTask,
    createToolSelectionLLMEvaluator
} from './evaluation-utils.js';
import {
    DATASET_NAME,
    MODELS_TO_EVALUATE,
    PASS_THRESHOLD,
    EVALUATOR_NAMES,
    type EvaluatorName,
    sanitizeHeaderValue,
    validateEnvVars
} from './config.js';

interface EvaluatorResult {
    model: string;
    experimentName: string;
    experimentId: string;
    evaluatorName: EvaluatorName;
    accuracy: number;
    correct: number;
    total: number;
    passed: boolean;
    error?: string;
}

/**
 * Interface for command line arguments
 */
interface CliArgs {
    datasetName?: string;
}

log.setLevel(log.LEVELS.DEBUG);

const RUN_LLM_EVALUATOR = true;
const RUN_TOOLS_EXACT_MATCH_EVALUATOR = true;

dotenv.config({ path: '.env' });

// Parse command line arguments using yargs
const argv = yargs(hideBin(process.argv))
    .wrap(null) // Disable automatic wrapping to avoid issues with long lines
    .usage('Usage: $0 [options]')
    .env()
    .option('dataset-name', {
        type: 'string',
        describe: 'Custom dataset name to evaluate (default: from config.ts)',
        example: 'my_custom_dataset',
    })
    .help('help')
    .alias('h', 'help')
    .version(false)
    .epilogue('Examples:')
    .epilogue('  $0                                    # Use default dataset from config')
    .epilogue('  $0 --dataset-name tmp-1               # Evaluate custom dataset')
    .epilogue('  npm run evals:run -- --dataset-name custom_v1  # Via npm script')
    .parseSync() as CliArgs;

// Sanitize secrets early to avoid invalid header characters in CI
process.env.OPENROUTER_API_KEY = sanitizeHeaderValue(process.env.OPENROUTER_API_KEY);

// Tools match evaluator: returns score 1 if expected tool_calls match output list, 0 otherwise
const toolsExactMatch = asEvaluator({
    name: EVALUATOR_NAMES.TOOLS_EXACT_MATCH,
    kind: 'CODE',
    evaluate: async ({ output, expected }: any) => {
        log.info(`Evaluating tools match. Expected: ${JSON.stringify(expected)}, Output: ${JSON.stringify(output)}`);

        let expectedTools = expected?.expectedTools || [];
        if (typeof expectedTools === 'string') {
            expectedTools = expectedTools.split(', ');
        }

        if (!expectedTools || expectedTools.length === 0) {
            log.debug('Tools match: No expected tools provided');
            return {
                score: 1.0,
                explanation: 'No expected tools present in the test case, either not required or not provided',
            };
        }

        // Normalize tool names: treat call-actor with step="info" as equivalent to fetch-actor-details
        const normalizeToolName = (toolName: string): string => {
            // Normalize call-actor to fetch-actor-details (bidirectional equivalence)
            if (toolName === 'call-actor' || toolName === 'fetch-actor-details') {
                return 'fetch-actor-details';
            }
            return toolName;
        };

        const normalizeToolCall = (toolCall: any): string => {
            const toolName = toolCall.function?.name || '';

            // If it's call-actor with step="info", treat it as fetch-actor-details
            if (toolName === 'call-actor') {
                try {
                    const args = JSON.parse(toolCall.function?.arguments || '{}');
                    if (args.step === 'info') {
                        return 'fetch-actor-details';
                    }
                } catch (e) {
                    // If we can't parse arguments, just return the tool name
                }
            }

            return toolName;
        };

        // Normalize expected tools (both call-actor and fetch-actor-details → fetch-actor-details)
        const normalizedExpectedTools = [...expectedTools]
            .map(normalizeToolName)
            .sort();

        const outputToolsTmp = (output?.tool_calls || [])
            .map(normalizeToolCall)
            .sort();

        const outputToolsSet = Array.from(new Set(outputToolsTmp)).sort();
        // it is correct if outputTools includes multiple calls to the same tool
        const isCorrect = JSON.stringify(normalizedExpectedTools) === JSON.stringify(outputToolsSet);
        const score = isCorrect ? 1.0 : 0.0;
        const explanation = `Expected: ${JSON.stringify(normalizedExpectedTools)}, Got: ${JSON.stringify(outputToolsSet)}`;

        log.debug(`🤖 Tools exact match: score=${score}, output=${JSON.stringify(outputToolsSet)}, expected=${JSON.stringify(normalizedExpectedTools)}`);

        return {
            score,
            explanation,
        };
    },
});


function processEvaluatorResult(
    experiment: any,
    modelName: string,
    experimentName: string,
    evaluatorName: EvaluatorName
): EvaluatorResult {
    const runsMap = experiment.runs ?? {};
    const evalRuns = experiment.evaluationRuns ?? [];
    const total = Object.keys(runsMap).length;

    const evaluatorRuns = evalRuns.filter((er: ExperimentEvaluationRun) => er.name === evaluatorName);
    const correct = evaluatorRuns.filter((er: ExperimentEvaluationRun) => (er.result?.score ?? 0) > 0.5).length;
    const accuracy = total > 0 ? correct / total : 0;

    return {
        model: modelName,
        experimentName,
        experimentId: experiment.id,
        evaluatorName,
        accuracy,
        correct,
        total,
        passed: accuracy >= PASS_THRESHOLD,
    };
}


function printResults(results: EvaluatorResult[]): void {
    log.info('📊 Results:');
    for (const result of results) {
        const { model, evaluatorName, accuracy, correct, total, passed, error } = result;
        if (error) {
            log.info(`${model}: ${evaluatorName} ❌ Error`);
        } else {
            const status = passed ? 'PASS' : 'FAIL';
            log.info(`${model}: ${evaluatorName} ${(accuracy * 100).toFixed(1)}% (${correct}/${total}) ${status}`);
        }
    }

    log.info(`\nPass threshold: ${(PASS_THRESHOLD * 100).toFixed(1)}%`);
    const allPassed = results.every(r => !r.error && r.passed);
    if (allPassed) {
        log.info('All tests passed');
    } else {
        log.info('Some tests failed');
    }
}

async function main(datasetName: string): Promise<number> {
    log.info('Starting MCP tool calling evaluation');

    if (!validateEnvVars()) {
        return 1;
    }

    const tools = await loadTools();
    log.info(`Loaded ${tools.length} tools`);

    // Phoenix client init (options may be provided via env)
    const client = createClient({
        options: {
            baseUrl: process.env.PHOENIX_BASE_URL!,
            headers: { Authorization: `Bearer ${sanitizeHeaderValue(process.env.PHOENIX_API_KEY)}` },
        },
    });

    // Resolve dataset by name -> id
    let datasetId: string | undefined;
    try {
        const info = await getDatasetInfo({ client, dataset: { datasetName } });
        datasetId = info?.id as string | undefined;
    } catch (e) {
        log.error(`Error loading dataset: ${e}`);
        return 1;
    }

    if (!datasetId) throw new Error(`Dataset "${datasetName}" not found`);

    log.info(`Loaded dataset "${datasetName}" with ID: ${datasetId}`);

    const results: EvaluatorResult[] = [];

    // Create the LLM evaluator with loaded tools
    const toolSelectionLLMEvaluator = createToolSelectionLLMEvaluator(tools);

    for (const modelName of MODELS_TO_EVALUATE) {
        log.info(`\nEvaluating model: ${modelName}`);

        // OpenRouter task
        const taskFn = createOpenRouterTask(modelName, tools);

        // Get PR info for better tracking
        const prNumber = process.env.GITHUB_PR_NUMBER || 'local';
        const prLabel = prNumber === 'local' ? 'local run' : `PR #${prNumber}`;

        const experimentName = `MCP server:  ${modelName}`;
        const experimentDescription = `${modelName}, ${prLabel}`;

        const evaluators = [];
        if (RUN_TOOLS_EXACT_MATCH_EVALUATOR) {
            evaluators.push(toolsExactMatch);
        }
        if (RUN_LLM_EVALUATOR) {
            evaluators.push(toolSelectionLLMEvaluator);
        }

        try {
            const experiment = await runExperiment({
                client,
                dataset: { datasetName },
                // Cast to satisfy ExperimentTask type
                task: taskFn as ExperimentTask,
                evaluators,
                experimentName,
                experimentDescription,
                concurrency: 10,
            });
            log.info(`Experiment run completed`);

            // Process each evaluator separately
            results.push(processEvaluatorResult(experiment, modelName, experimentName, EVALUATOR_NAMES.TOOLS_EXACT_MATCH));
            results.push(processEvaluatorResult(experiment, modelName, experimentName, EVALUATOR_NAMES.TOOL_SELECTION_LLM));
        } catch (e: unknown) {
            const err = e instanceof Error ? e : new Error(String(e));
            log.error(`Error evaluating ${modelName}:`, err);
            log.error(`Full error trace: ${err.stack ?? err.message}`);

            // Add error results for both evaluators
            Object.values(EVALUATOR_NAMES).forEach(evaluatorName => {
                results.push({
                    model: modelName,
                    experimentName,
                    experimentId: '',
                    evaluatorName,
                    accuracy: 0,
                    correct: 0,
                    total: 0,
                    passed: false,
                    error: err.message
                });
            });
        }
    }

    printResults(results);

    const allPassed = results.every(r => !r.error && r.passed);
    return allPassed ? 0 : 1;
}

// Run
main(argv.datasetName || DATASET_NAME)
    .then((code) => process.exit(code))
    .catch((err) => {
        log.error('Unexpected error:', err);
        process.exit(1);
    });
